[build.webkit.org] Switch macOS Debug JSC stress test queue to O3 debug builds

ryanhaddad opened the pull request:
#### c658e38c091d211f78c30a83421aaea6abe41e0b
<pre>
[build.webkit.org] Switch macOS Debug JSC stress test queue to O3 debug builds
<a href="https://bugs.webkit.org/show_bug.cgi?id=302031">https://bugs.webkit.org/show_bug.cgi?id=302031</a>
<a href="https://rdar.apple.com/159975242">rdar://159975242</a>

Reviewed by Jonathan Bedard.

Switching this queue to O3 debug builds will increase its throughput and test result granularity.
This change brings test runtime down from 11 hours to about an hour on the same hardware. If this
queue proves to be reliable, we may switch the JSC-Tests-arm64-EWS queue to this configuration in
order to catch assertion failures on JSC pull requests.

In order to limit the scope of O3 builds to JSC tests, this test queue must now compile JSC
itself rather than being triggered by the same debug builder that triggers all macOS debug
style tests.

* Tools/CISupport/build-webkit-org/config.json: Update the queue name to reflect that it is
now a build and test queue, and switch it to the new factory.
* Tools/CISupport/build-webkit-org/factories.py:
(BuildO3AndJSCTestsFactory): Create a new factory that will include the step to set the
optimization level.
(BuildO3AndJSCTestsFactory.__init__):
* Tools/CISupport/build-webkit-org/factories_unittest.py:
(TestExpectedBuildSteps): Update unit tests.
* Tools/CISupport/build-webkit-org/public_html/dashboard/Scripts/WebKitBuildbot.js:
(WebKitBuildbot): Update the queue name on the dashboard.
* Tools/CISupport/build-webkit-org/steps.py:
(SetO3OptimizationLevel): New step to run `set-webkit-configuration` to enable the O3 optimization
level before building JSC.

Canonical link: <a href="https://commits.webkit.org/302930@main">https://commits.webkit.org/302930@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/a25543b6e69c95d3840acda530bc9a9f7bb783c3

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/130672 "Passed style check") | [  ~~🛠 ios~~](https://ews-build.webkit.org/#/builders/159/builds/2943 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/41627 "Build is in progress. Recent messages:") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/5/builds/138095 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/82301 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/f731876e-a3bb-44ce-9a25-27b557f22865) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/132543 "Passed tests") | [  ~~🛠 ios-sim~~](https://ews-build.webkit.org/#/builders/155/builds/2959 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 mac-AS-debug~~](https://ews-build.webkit.org/#/builders/156/builds/2836 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/5/builds/138095 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/82301 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/dfdfd9b0-3723-4265-8254-ce600497fcaf) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/133619 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/155/builds/2959 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 api-mac](https://ews-build.webkit.org/#/builders/138/builds/41627 "Build is in progress. Recent messages:") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/5/builds/138095 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/4bdd0be7-1c72-42e5-941b-2f37993365f2) 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/155/builds/2959 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/138/builds/41627 "Build is in progress. Recent messages:") | [  ~~🛠 wpe-cairo~~](https://ews-build.webkit.org/#/builders/65/builds/81349 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/155/builds/2959 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/138/builds/41627 "Build is in progress. Recent messages:") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/2/builds/140573 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | | 
| | [  ~~🛠 vision~~](https://ews-build.webkit.org/#/builders/153/builds/2733 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/156/builds/2836 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/2/builds/140573 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | | 
| [✅ 🧪 services](https://ews-build.webkit.org/#/builders/28/builds/130102 "Passed tests") | [  ~~🛠 vision-sim~~](https://ews-build.webkit.org/#/builders/160/builds/2777 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/138/builds/41627 "Build is in progress. Recent messages:") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/2/builds/140573 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | | 
| | [⏳ 🧪 vision-wk2 ](https://ews-build.webkit.org/#/builders/visionOS-26-Simulator-WK2-Tests-EWS "Waiting in queue, processing has not started yet") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/138/builds/41627 "Build is in progress. Recent messages:") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/55726 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/20345 "Built successfully and passed tests") | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/158/builds/2803 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/157/builds/2623 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/163/builds/2824 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/152/builds/2729 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | | 
<!--EWS-Status-Bubble-End-->

bls1999 commented on Tools/CISupport/build-webkit-org/public_html/dashboard/Scripts/WebKitBuildbot.js:
```diff
@@ -68,7 +68,7 @@ WebKitBuildbot = function()
             "Apple-Sonoma-Debug-Test262-Tests": {heading: "Debug Test262 (Tests)"},
             "Apple-Sonoma-Release-Test262-Tests": {heading: "Release Test262 (Tests)"},
-            "Apple-Sonoma-AppleSilicon-Debug-JSC-Tests": {heading: "Debug arm64 JSC (Tests)"},
+            "Apple-Sonoma-AppleSilicon-Debug-JSC-BuildAndTest": {heading: "Debug arm64 JSC (Tests)"},
```
The end of the heading on this should change, as this is no longer only a test queue. Maybe something like "O3 build+test"?

bls1999 reviewed (approved):
1 small comment, otherwise [unofficial] r=me

ryanhaddad committed:
[build.webkit.org] Switch macOS Debug JSC stress test queue to O3 debug builds
https://bugs.webkit.org/show_bug.cgi?id=302031
rdar://159975242

Reviewed by Jonathan Bedard.

Switching this queue to O3 debug builds will increase its throughput and test result granularity.
This change brings test runtime down from 11 hours to about an hour on the same hardware. If this
queue proves to be reliable, we may switch the JSC-Tests-arm64-EWS queue to this configuration in
order to catch assertion failures on JSC pull requests.

In order to limit the scope of O3 builds to JSC tests, this test queue must now compile JSC
itself rather than being triggered by the same debug builder that triggers all macOS debug
style tests.

* Tools/CISupport/build-webkit-org/config.json: Update the queue name to reflect that it is
now a build and test queue, and switch it to the new factory.
* Tools/CISupport/build-webkit-org/factories.py:
(BuildO3AndJSCTestsFactory): Create a new factory that will include the step to set the
optimization level.
(BuildO3AndJSCTestsFactory.__init__):
* Tools/CISupport/build-webkit-org/factories_unittest.py:
(TestExpectedBuildSteps): Update unit tests.
* Tools/CISupport/build-webkit-org/public_html/dashboard/Scripts/WebKitBuildbot.js:
(WebKitBuildbot): Update the queue name on the dashboard.
* Tools/CISupport/build-webkit-org/steps.py:
(SetO3OptimizationLevel): New step to run `set-webkit-configuration` to enable the O3 optimization
level before building JSC.

Canonical link: https://commits.webkit.org/302930@main
